monitor/services/src/main/java/org/datacleaner/monitor/configuration/ComponentsCache.java

```diff
@@ -0,0 +1,233 @@
+            }
+        }
+        componentsCacheConfigWrapper.updateTimeStamp();
+        return componentsCacheConfigWrapper;
+    }
+
+    /**
+     * Remove configuration from memory and store. And this component is destroyed.
+     *
+     * @param id
+     * @param tenantContext
+     */
+    public boolean removeConfiguration(String id, TenantContext tenantContext) {
+        boolean inCache = removeConfigurationOnlyFromCache(id);
+        boolean inStore = removeConfigurationOnlyFromStore(id, tenantContext);
+        logger.info("Component {} was removed from cache and closed.", id);
+        return inCache || inStore;
+    }
+
+
+    private boolean removeConfigurationOnlyFromCache(String id) {
+        ComponentsCacheConfigWrapper config = data.get(id);
+        if (config != null) {
+            data.remove(id);
+            config.getHandler().closeComponent();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean removeConfigurationOnlyFromStore(String id, TenantContext tenantContext) {
+        ComponentStore store = tenantContext.getComponentsStore();
+        return store.removeConfiguration(id);
+    }
+
+    /**
+     * Close all component in memory. All components configuration are still in repository.
+     *
+     * @throws InterruptedException
+     */
+    public void close() throws InterruptedException {
+        logger.info("Closing Components cache.");
+
+        synchronized (checker) {
+            checker.stop();
+            checker.notifyAll();
+        }
+        long maxTime = System.currentTimeMillis() + CLOSE_TIMEOUT;
+        while (checkerThread.isAlive()) {
+            Thread.sleep(500);
+            if (maxTime < System.currentTimeMillis()) {
+                logger.error("Problem with closing checking thread.");
+                break;
+            }
+        }
+
+        for (ComponentsCacheConfigWrapper componentsCacheConfigWrapper : data.values()) {
+            componentsCacheConfigWrapper.getHandler().closeComponent();
+            logger.info("Component with id: {} was closed.", componentsCacheConfigWrapper.getComponentsStoreHolder().getComponentId());
+            // Configuration is still in store.
+        }
+
+        data.clear();
+        logger.info("Components cache was closed.");
+    }
+
+
+    /**
+     * Thread for checking timeout for each components and also do update configuration is store.
+     */
+    private class TimeoutChecker implements Runnable {
+        boolean running = true;
+        boolean firstRun = true;
+
+        @Override
+        public void run() {
+            while (running) {
+                if (firstRun) {
+                    firstRun = false;
+                } else {
+                    Set<String> tenants = _tenantContextFactory.getAllTenantsName();
+                    Set<String> allIdInCache = new HashSet<>(data.keySet());
+                    for (String tenant : tenants) {
```

Comment:
Ouch, this is not a good approach... We should cache this in the TenantContexts that are active only. Here we are iterating and calling getContext(tenant) on _all_ the tenants. That's potentially a very big iteration and calling getContext(tenant) is very expensive to do in cases where there is no need for the context.